HeterogeneousCore/CUDAUtilities/interface/radixSort.h

```diff
@@ -0,0 +1,266 @@
+  //move negative first...
+
+  int32_t first = threadIdx.x;
+  __shared__ uint32_t firstNeg;
+  firstNeg = a[ind[0]] < 0 ? 0 : size;
+  __syncthreads();
+
+  // find first negative
+  for (auto i = first; i < size - 1; i += blockDim.x) {
+    if ((a[ind[i]] ^ a[ind[i + 1]]) < 0)
+      firstNeg = i + 1;
+  }
+
+  __syncthreads();
+
+  auto ii = first;
+  for (auto i = firstNeg + threadIdx.x; i < size; i += blockDim.x) {
+    ind2[ii] = ind[i];
+    ii += blockDim.x;
+  }
+  __syncthreads();
+  ii = size - firstNeg + threadIdx.x;
+  assert(ii >= 0);
+  for (auto i = first; i < firstNeg; i += blockDim.x) {
+    ind2[ii] = ind[i];
+    ii += blockDim.x;
+  }
+  __syncthreads();
+  for (auto i = first; i < size; i += blockDim.x)
+    ind[i] = ind2[i];
+}
+
+template <typename T>
+__device__ inline void reorderFloat(T const* a, uint16_t* ind, uint16_t* ind2, uint32_t size) {
+  //move negative first...
+
+  int32_t first = threadIdx.x;
+  __shared__ uint32_t firstNeg;
+  firstNeg = a[ind[0]] < 0 ? 0 : size;
+  __syncthreads();
+
+  // find first negative
+  for (auto i = first; i < size - 1; i += blockDim.x) {
+    if ((a[ind[i]] ^ a[ind[i + 1]]) < 0)
+      firstNeg = i + 1;
+  }
+
+  __syncthreads();
+
+  int ii = size - firstNeg - threadIdx.x - 1;
+  for (auto i = firstNeg + threadIdx.x; i < size; i += blockDim.x) {
+    ind2[ii] = ind[i];
+    ii -= blockDim.x;
+  }
+  __syncthreads();
+  ii = size - firstNeg + threadIdx.x;
+  assert(ii >= 0);
+  for (auto i = first; i < firstNeg; i += blockDim.x) {
+    ind2[ii] = ind[i];
+    ii += blockDim.x;
+  }
+  __syncthreads();
+  for (auto i = first; i < size; i += blockDim.x)
+    ind[i] = ind2[i];
+}
+
+template <typename T,  // shall be interger
+          int NS,      // number of significant bytes to use in sorting
+          typename RF>
+__device__ void __forceinline__
+radixSortImpl(T const* __restrict__ a, uint16_t* ind, uint16_t* ind2, uint32_t size, RF reorder) {
+  constexpr int d = 8, w = 8 * sizeof(T);
+  constexpr int sb = 1 << d;
+  constexpr int ps = int(sizeof(T)) - NS;
+
+  __shared__ int32_t c[sb], ct[sb], cu[sb];
+
+  __shared__ int ibs;
+  __shared__ int p;
+
+  assert(size > 0);
+  assert(blockDim.x >= sb);
+
+  // bool debug = false; // threadIdx.x==0 && blockIdx.x==5;
+
+  p = ps;
+
+  auto j = ind;
+  auto k = ind2;
+
+  int32_t first = threadIdx.x;
+  for (auto i = first; i < size; i += blockDim.x)
+    j[i] = i;
+  __syncthreads();
+
+  while (__syncthreads_and(p < w / d)) {
+    if (threadIdx.x < sb)
+      c[threadIdx.x] = 0;
+    __syncthreads();
+
+    // fill bins
+    for (auto i = first; i < size; i += blockDim.x) {
+      auto bin = (a[j[i]] >> d * p) & (sb - 1);
+      atomicAdd(&c[bin], 1);
+    }
+    __syncthreads();
+
+    // prefix scan "optimized"???...
+    if (threadIdx.x < sb) {
+      auto x = c[threadIdx.x];
+      auto laneId = threadIdx.x & 0x1f;
+#pragma unroll
```

Comment:
The header consistency tool issues a warning on unknown pragma (I remember gcc or clang issuing a warning as well elsewhere). I'm not sure what to do though.

Reply:
I'll leave this for the moment.